querybook/webapp/ui/RichTextEditor/RichTextEditor.tsx

```diff
@@ -45,523 +55,562 @@ export interface IRichTextEditorState {
+    const toolBarWidth = element.clientWidth;
 
-    public set editorState(editorState: DraftJs.EditorState) {
-        this.setState({
-            editorState,
-        });
+    const selectionRect = Utils.getSelectionRect();
+    if (selectionRect == null) {
+        return null;
     }
 
-    @bind
-    public focus() {
-        if (this.editorRef) {
-            this.editorRef.current.focus();
-            this.setState({
-                toolBarStyle: this.calculateToolBarStyle(
-                    this.state.editorState.getSelection()
-                ),
-            });
-        }
-    }
+    const position = {
+        top: Math.max(selectionRect.top - toolBarHeight - 5, 0),
+        left: Math.max(
+            selectionRect.left + (selectionRect.width - toolBarWidth) / 2,
+            0
+        ),
+    };
+    return position;
+}
 
-    @bind
-    public calculateToolBarPosition() {
-        if (!this.toolBarRef.current?.selfRef) {
-            return null;
-        }
-        const toolBarDivRef = this.toolBarRef.current.selfRef;
-
-        const toolBarHeight = toolBarDivRef.current.clientHeight;
-        const toolBarWidth = toolBarDivRef.current.clientWidth;
-
-        const selectionRect = Utils.getSelectionRect();
-        if (selectionRect == null) {
-            return null;
-        }
-
-        const position = {
-            top: Math.max(selectionRect.top - toolBarHeight - 5, 0),
-            left: Math.max(
-                selectionRect.left + (selectionRect.width - toolBarWidth) / 2,
-                0
-            ),
-        };
-        return position;
-    }
+function calculateToolBarStyle(
+    selection: DraftJs.SelectionState,
+    toolBarDiv: Nullable<HTMLElement>
+): IToolBarStyle {
+    const isVisible = !selection.isCollapsed() && selection.getHasFocus();
+    const position =
+        (isVisible ? calculateToolBarPosition(toolBarDiv) : null) || {};
+
+    return {
+        ...position,
+        visibility: (isVisible ? 'visible' : 'hidden') as VisibilityProperty,
+    };
+}
 
-    @bind
-    public calculateToolBarStyle(selection: DraftJs.SelectionState) {
-        const isVisible = !selection.isCollapsed() && selection.getHasFocus();
-        const position =
-            (isVisible ? this.calculateToolBarPosition() : null) || {};
-
-        return {
-            ...position,
-            visibility: (isVisible
-                ? 'visible'
-                : 'hidden') as VisibilityProperty,
-        };
-    }
+export const RichTextEditor = React.forwardRef<
+    IRichTextEditorHandles,
+    IRichTextEditorProps
+>(
+    (
+        {
+            className,
+            readOnly = false,
+            value,
+            placeholder,
+            autoFocus,
+            decorator,
+
+            onChange,
+            onKeyDown,
+            onFocus,
+            onBlur,
+        },
+        ref
+    ) => {
+        const [editorState, setEditorState, editorStateRef] = useStateWithRef(
```

Comment:
dont quite get why we need `useStateWithRef` here.
1. I dont see `editorState` is used in any `useEffect`'s deps array.
2. If we'd like to not trigger useEffect update when editorState changes, why not just use `useRef` instead of state?

Reply:
its to suppress frequent updates in useImperativeHandles and useCallback